x-pack/plugins/apm/public/components/app/service_groups/service_groups_list/service_groups_list.tsx

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { EuiFlexGrid } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import React from 'react';
+import { useHistory } from 'react-router-dom';
+import { TypeOf } from '@kbn/typed-react-router-config';
+import { SavedServiceGroup } from '../../../../../common/service_groups';
+import { ServiceGroupsCard } from './service_group_card';
+import { SERVICE_GROUP_COLOR_DEFAULT } from '../../../../../common/service_groups';
+import { useApmRouter } from '../../../../hooks/use_apm_router';
+import { ApmRoutes } from '../../../routing/apm_route_config';
+
+interface Props {
+  items: SavedServiceGroup[];
+  isLoading: boolean;
+}
+
+export function ServiceGroupsListItems({ items }: Props) {
+  const router = useApmRouter();
+  const history = useHistory();
+  const { query } = router.getParams('/*', history.location);
+  return (
+    <EuiFlexGrid gutterSize="m">
+      {items.map((item) => (
+        <ServiceGroupsCard
+          serviceGroup={item}
+          href={router.link(
+            '/services',
+            {
+              query: { ...query, serviceGroup: item.id } as TypeOf<
+                ApmRoutes,
+                '/services'
+              >['query'],
+            },
+            true // skips runtime param validation
```

Comment:
I don't understand why this is necesssary - and I think making validation optional is not the right thing to do here. Do you mind describing what issues you're running into?

Reply:
i want to maintain all query params except for `serviceGroup` when the user navigates to `/service-groups`, so that their filters are not reset. I want to do this without having to define every possible query param in the `/service-groups` route. Earlier, i did this by just using history.location.search, but in feedback was encouraged to use `router.link`, which validates at runtime each param. So i added this option so i can still use `router.link` while passing every query param thru to the next service group.

Reply:
just here

Reply:
These are the various approaches i've tried to accomplish what i want:

```typescript
  // Example URL:
  //    /app/apm/service-groups?
  //    comparisonEnabled=true&comparisonType=day&environment=production&kuery=&rangeFrom=now-30m&rangeTo=now&serviceGroup=
```

Idiomatic way to get query params, does not include everything (e.g. environment):
```typescript
  const { query } = useApmParams('/service-groups'); // same as '/*'
  // query={ rangeFrom: 'now-30m', rangeTo: 'now', serviceGroup: '' }
```

My current approach here, similarly is not working as intended:
```typescript
  const history = useHistory();
  const { query } = router.getParams('/*', history.location);
  // query={ rangeFrom: 'now-30m', rangeTo: 'now', serviceGroup: '' }
```

However this approach (not in branch) is the only one that works as intended:
```typescript
  const { query } = router.getParams('/*', { ...window.location, state: undefined });
  // query={ environment: 'production', kuery: '', rangeFrom: 'now-30m', rangeTo: 'now', comparisonEnabled: true, comparisonType: 'day' }
```

This is the original approach i used in the first pass, which works mostly as intended, however all values are always strings.
```typescript
  const history = useHistory();
  const query = qs.parse(history.location.search);
  // query={ environment: 'production', kuery: '', rangeFrom: 'now-30m', rangeTo: 'now', comparisonEnabled: 'true', comparisonType: 'day', serviceGroup: '' }
```

The ultimate goal here is to allow the user who is using and filtering APM to navigate to back to select a new service group and maintain their filters without them being reset.

Reply:
We resolved to address this by defining date range params in the `/service-groups` route in order to use `useApmParams` to pass them along to `router.link`. The other params can be reset with default values. This prevents me from having to side-step runtime query validation or use an off-pattern approach to obtaining query params. Thanks for the help @dgieselaar!